### ISSUE TITLE
How to exclude `type` imports from being processed by `forbidden` rule

### ISSUE THREAD

**Issue by agomai** [sverweij/dependency-cruiser#490]
### Summary

Is there a way to exclude TypeScript `type` only imports from being considered in a `forbidden` rule?

### Context

I have a rule where I restrict a module from importing anything but some allowed modules, but I want that it is still allowed to import `type` definitions from these modules.

In my specific case I want a skeleton component only to be allowed to import from some allow-listed modules to prevent bundle size bloats. Importing types should be allowed as these don't affect bundle size

~~~javascript
{
    name: 'skeleton-only',
    severity: 'error',
    from: {
      path: ['.+-skeleton\\.tsx$'],
    },
    to: {
      dependencyTypes: ['local', 'localmodule'],
     // is this the way?
     // preCompilationOnly: true,

      // forbid all…
      path: [`^src/.*`],
      // …allow:
      pathNot: [
        // other skeletons
        `.+-skeleton.*\\.tsx$`,
       // others excluded for brevity
      ],
    },
  }
~~~

Currently I configured `options` `tsPreCompilationDeps: true` as I generally want Typescript dependencies to be considered.
I tried `tsPreCompilationDeps: specify` with `preCompilationOnly: true` in the rule and a type import from a forbidden module was not considered (the desired behavior). However this was also the case with `preCompilationOnly: false`, so I'm not sure if this works right. Setting `preCompilationOnly: undefined` caused no violations to be found at all (both type and regular imports) for the rule (bug?). Only not setting `preCompilationOnly` at all caused the `type` imports to be causing errors again  – not the desired behavior, but I tried as a negative test and to find out what `tsPreCompilationDeps: specify` does by default as I want all other rules to still consider type imports.

So questions are: 

1. How do I exclude TypeScript `type` only imports from being considered in a `forbidden` rule?
  * Is there are feature to explicitly target `type` definitions which I can enable/disable per rule or do I have to use `tsPreCompilationDeps` with `preCompilationOnly`
2. If `tsPreCompilationDeps` with `preCompilationOnly` is the way:
  * Which config do I have to use to only disable considering `type` imports for that rule but still keep other rules to consider TypeScript
  * What is the default behavior for `tsPreCompilationDeps: specify`? Does a rule consider TypeScript by default if `preCompilationOnly` is not set?
  * Why does `preCompilationOnly: true/false` seem behave the same way?
  * Why does `preCompilationOnly: undefined` caused no violations to be found at all, is this a bug?

Thanks for you help and your awesome work on dependency cruiser 👍

**Comment by sverweij** [comment]
hi @agomai thanks for asking! 

### addressing your questions ...

> How do I exclude TypeScript type only imports from being considered in a forbidden rule?

It seems you have all elements necessary to get what you want:
- set `tsPreCompilationDeps` in `options` to `"specify"`
- use the `preCompilationOnly` attribute in the rule to have it look at dependencies that exist only before compilation (e.g to modules that are 100% types, or that are imported but not used) or the other way 'round.

So approximately like in this (very minimal) [.dependency-cruiser.js](https://github.com/sverweij/dependency-cruiser-repro-repo/blob/main/490/.dependency-cruiser.js)

> Is there are feature to explicitly target type definitions which I can enable/disable per rule or do I have to use tsPreCompilationDeps with preCompilationOnly?

Currently that is indeed the route. I can (and probably should) add detection for [type only imports and exports](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-8.html#type-only-imports-and-export) that are explicitly set (`import type { ThingType } from 'respond'`) and add that to the detected `dependencyTypes` so you can also use that in rules. For implicit type imports (the equally legal `import { ThingType } from 'respond'`) that won't help, though and the preCompilationOnly attribute will remain the only option I see a.t.m.


> Which config do I have to use to only disable considering type imports for that rule but still keep other rules to consider TypeScript

See above


> What is the default behavior for tsPreCompilationDeps: specify? Does a rule consider TypeScript by default if preCompilationOnly is not set?

When not set in the rule, the rule should match  _both_ things that exist only before compilation and those that make it through compilation.

> Why does preCompilationOnly: true/false seem behave the same way?

That shouldn't be the case, at least, not when `tsPreCompilationDeps` equals `"specify"`. I can't reproduce the problem - but very interested in a case that 

> Why does preCompilationOnly: undefined caused no violations to be found at all, is this a bug?

That is indeed unexpected behaviour. I'll add a unit test for it and mend (or worst case document) the behaviour.

### reproduction example

I've made a minimal reproduction example in the [repro repo](https://github.com/sverweij/dependency-cruiser-repro-repo/tree/main/490) to make sure it still works. The [configuration file](https://github.com/sverweij/dependency-cruiser-repro-repo/blob/main/490/.dependency-cruiser.js) has three rules: 
- one that is only applicable for modules that exist both before and after compilation ([_dont-depend-on-code_](https://github.com/sverweij/dependency-cruiser-repro-repo/blob/main/490/.dependency-cruiser.js#L5))
- one that is applicable to modules that exist _only_ before compilation ([_dont-depend-on-pre-com-only_](https://github.com/sverweij/dependency-cruiser-repro-repo/blob/main/490/.dependency-cruiser.js#L16))
- one applicable to modules regardless whether they exist before compilation only or also after

They seem to work as expected. In your case, they don't though, so there's bound to be a difference between this repro sample and what is in your config/ code base.  It could be that a combination preCompilationOnly and other attributes triggers a specific bug. Or maybe there's something specific with the type-only modules in your repo (e.g. it has something the typescript compiler deems useful enough to keep through the compilation process). If the above doesn't help - could you provide a minimal reproduction example so I can investigate what's going on?

**Comment by agomai** [comment]
Thanks @sverweij for your detailed answer.

I was able to get it working 👍🙏

> Currently that is indeed the route. I can (and probably should) add detection for type only imports and exports that are explicitly set (import type { ThingType } from 'respond') and add that to the detected dependencyTypes so you can also use that in rules.

That would be very helpful as we are only using `type` only imports (enforced via lint rules). This way we can further refine the dependency cruiser rules (e.g. regarding bundle size, cyclic dependencies…).

>> Why does preCompilationOnly: true/false seem behave the same way?

> That shouldn't be the case, at least, not when tsPreCompilationDeps equals "specify". I can't reproduce the problem - but very interested in a case that does.

I checked again and couldn't reproduce it either. Probably was a misconfiguration on my side while checking all the combinations.

Thanks again, I'm excited to see the next releases of dependency cruiser 😀

**Comment by sverweij** [comment]
Hi @agomai  - I've added support for 'type only' imports (as MicroSoft's typescript compiler calls them) - see the PR linked above for details. From now it's possible to use `type-only` in rules (next to things like `localmodule`, `npm`, `core` etc) when imports are done explicitly as a type import (e.g. `import type { IYadda} from "yadda"`). 

I've published it as `dependency-cruiser@10.4.0-beta-1` - if you want to give it a swing; I'm curious about any feedback you'd have.
